### PR TITLE
Implementing Support Stamina Weapon/Character Skills

### PR DESCRIPTION
**Pull request by OrkunKocyigit**
This pull change aims to bring special support stamina skill that following characters have:
----
- weapon skill
  - [x] [Cherubim lance](https://gbf.wiki/Cherubim%27s_Lance) - %10 Normal Stamina
  - [x] [True Purity Sunblade](https://gbf.wiki/True_Purity_Sunblade) - %15 Normal Stamina
- stamina support ability
  - [x] [Arulumaya (Summer)](https://gbf.wiki/Arulumaya_(Summer)) - %15 Normal Stamina
  - [x] [Izmir (Summer)](https://gbf.wiki/Izmir (Summer)) - %15 Normal Stamina
  - [x] [Jeanne d'Arc](https://gbf.wiki/Jeanne_d'Arc) - %15 Normal Stamina
  - [x] [Sakura Kinomoto](https://gbf.wiki/Sakura%20Kinomoto) - %15 Normal Stamina
----
This patch adds to:
- [x] normalSupportKonshin for characters.
- [x] normalSupportKonshinWeapon for weapons.
- [x] calculates and adds maximum of normalSupportKonshin and normalSupportKonshinWeapon and adds to normalKonshin additionally.
- [x] provides function wrapper for chart.js global_logic.js/recalcNormalSupportKonshin

### CODE REVIEW

**Comment by OrkunKocyigit** [comment]
I don't like current solution to buffs that effects regardless of elements. I'll think another solution.

**Comment by h-yasha** [comment]
The skill boosts all Allies' **ATK**.
`hissatsu` would be referring to refers to 必殺 (sentence) skills-or alike weapon skills which boosts C.A. DMG and C.A. DMG Cap.
Also This skill is also is on:
[True Purity Sunblade](https://gbf.wiki/True_Purity_Sunblade), and as support ability for characters. So, it would be better to add it to `calcHaisuiValue` function.

**Comment by OrkunKocyigit** [comment]
Oh I don't know japanese so I copied key name from huanlong spear. I just checked ixaba for reference so I think correct key name should be Konshin. I'll update key and push it.

**Comment by OrkunKocyigit** [comment]
> [True Purity Sunblade](https://gbf.wiki/True_Purity_Sunblade), and as support ability for characters. So, it would be better to add it to `calcHaisuiValue` function.

I am assuming you reference this logic right:
`let bonus = totals[key]["remainHP"] < 0.50 ? 0 : 18 * totals[key]["remainHP"] - 8;`

I'll see if it can be moved to that function instead of doing it in that if.

**Comment by h-yasha** [comment]
Note:
on this weapon it's different from (weaker compared to) the other ones (which all appear to be the same). Those are: True Purity Sunblade 2nd skill, and the support ablity of Arulumaya (Summer), Izmir (Summer), Jeanne d'Arc and Sakura Kinomoto.

**Comment by OrkunKocyigit** [comment]
I moved logic to function and call the function instead. From what I see motocal doesn't implement this logic other weapon/chars in the game. I'll check those you link.

**Comment by OrkunKocyigit** [comment]
> **Update:** You could simply add the weapon skill outside the element check (as `else if` before the `else if ` _`(elementcheck)`_)
I thought about that too but skill 'all' option doesn't require any change on code where as writing else if in those cases will require our code size to increase constantly. So my solution sounded more optimal.
> Also, the skill is based on Djeeta's HP only.
Thanks, fixed in 669207a

**Comment by h-yasha** [comment]
Refer to [this](https://github.com/MotocalDevelopers/motocal#updating-weapon-templates) about onto how to change templates appropriately.

Omega (ultima)'s strength key (stamina) is also calculated on `normalOtherKonshin`. This could conflict with it .I do not have verification (did not find any) for this, whether they do stack or not (I'll try to test it out).

**Comment by h-yasha** [comment]
image
image

Appears as that it does stacks.

**Comment by OrkunKocyigit** [comment]
> Refer to [this](https://github.com/MotocalDevelopers/motocal#updating-weapon-templates) about onto how to change templates appropriately.

Updated for weapons in 64fd829

**Comment by kei-gbf** [comment]
Can you check "emnity_all_SL10" support ability, it's a similar support ability, Haisui version.
I think "normalOtherKonshin" is a skill for weapon (not chara)

Naming is mix en/ja here but those skill/support id are stored in save data.
we can't rename them easily.

**Comment by OrkunKocyigit** [comment]
@kei-gbf 
Okay I am confused about char passives.
Our current code works like this:
1) Creates initial totals with all selection from ui and assign it to totals with getInitialTotals()
2) Adds character support skills to totals treatSupportAbility
3) Calculates result for 0 case (0, 0, 0, 0) with totals.
4) Then calls a method called another initializeTotals that sets every buff value to 0.

If I am not missing something then basically unless you do 0 case character buffs are never used. Is it a bug?

**Comment by OrkunKocyigit** [comment]
Edit: I am a fucking moron. I'll refactor the code to not use same skillset

**Comment by h-yasha** [comment]
~~woops wrote in the wrong textbox~~
> Can you check "emnity_all_SL10" support ability, it's a similar support ability, Haisui version.

There's a small difference however, enmity support abilities are unique unlike stamina support abilities (or sunblade/cherubim MH skill) is that it is **additive** with normal stamina.

> I think "normalOtherKonshin" is a skill for weapon (not chara)
> 
> Naming is mix en/ja here but those skill/support id are stored in save data.
> we can't rename them easily.

Should probably introduce something, like `normalOtherKonshin2` or something, as currently there is no 'variable' to hold this skill/support ability. Since `normalOtherKonshin` is used by omega strength weapon skill, although there is a weapon skill declared, `normalOtherKonshinL`, no actual weapon is using it currently.

File: src/global_logic.js [hunk 1]
```diff
@@ -2473,6 +2473,16 @@ module.exports.treatSupportAbility = function (totals, chara) {
                         totals[key]["ougiDamageLimitBuff"] += 0.10;
                     }
                     continue;
+                case "normalOtherKonshin_15UP":
+                    let otherHaisuiValue = module.exports.calcHaisuiValue("normalOtherKonshin", "L", 1, totals[key]["remainHP"]);
```
[the remark this answers is not in the document]
Comment (OrkunKocyigit) [reply src/global_logic.js]: Fixed in a314896

**Comment by OrkunKocyigit** [comment]
@h-yasha Latest version should fix most if not all the issues.

**Comment by kei-gbf** [comment]
@h-yasha 
> enmity support abilities are unique unlike stamina support abilities (or sunblade/cherubim MH skill) is that it is additive with normal stamina.

add skill value to `normalOtherKonshin` is ok, it's an Intermediate field which not in save-data.
about support ability name/id,

Updated chara_data_converter.py
```python
supportAbilist["normalOtherKonshin_15UP"] = {
    u"夏祭りの思い出",
    u"白の切っ先",
    u"不退転の戦旗",
    u"絶対だいじょうぶだよ"
}
```

At the First, I though was "stamina_all_xxx" for them.
there are similar support ability already implemented.

- emnity_all_SL10
- emnity_own_SL20

two points I wanted to mention here, the target "all", "own"
and "15UP" is a maximum value(when reminaHP 100%) it's not fixed value.

The difference as you say (additive with normal stamina) is also another additional info.
so "normal_stamina_all_SLxx" or any order ok, suffix SLxx_normal etc.

they will be stored in save-data, this is a research for existing support ability names.
I want there to have something similar naming rule.

**Comment by h-yasha** [comment]
> add skill value to normalOtherKonshin is ok, it's an Intermediate field which not in save-data.

but then it'd be hard to maintain that this effect does not stack with others similar to it (and only strongest take effect).

`stamina_all_L_hp_down_own_15` will not be calculated (HP cut) correctly if ever. I think having the effect of this split is fine. most characters doesn't actually get 2 or 3 support skills that really impact. (therefore, Character Innate C.A. DMG Cap is already implemented as a support ability)

**Comment by kei-gbf** [comment]
@h-yasha 
- only strongest take effect

I was missing this point.

my point is just support ability name/id, and there is similar one emnity_all_SL10
supportAbilist["normalOtherKonshin_15UP"]

**Comment by OrkunKocyigit** [comment]
> `stamina_all_L_hp_down_own_15`

I just replicated current code behavior as it is. Do you have any advice on it?

**Comment by h-yasha** [comment]
@OrkunKocyigit check [PR](https://github.com/OrkunKocyigit/motocal/pull/2)

**Comment by kei-gbf** [comment]
stamina_all_15up  or something

I don't know the skill amount of the stamina, no idea.
first I thought if it was fixed value ... 
but no chance to be fixed value it's stamina, change by remainHP.
so it's ok just max amount value.

Other developers who read Motocal code in future,
may assume if there is emnity_ ... and then stamina_ support ability.
no need to be so long? I don't think so many stamina_all variations are implement.
at least now 4 characters reuse the same support.

One more common naming rules I found in Motocal (but not all)
- skill name has camelCase
- support name are joined by "_" underscore.

**Comment by OrkunKocyigit** [comment]
@h-yasha Thanks for your help, I checked last commit and couldn't see anything wrong with it.

**Comment by OrkunKocyigit** [comment]
Oki I found a bug, it adds the weapon skill whatever it is equipped or not, I'll check on it
image

**Comment by OrkunKocyigit** [comment]
Fixed in 6ff75f0

**Comment by h-yasha** [comment]
> Fixed in [426c8a3](https://github.com/MotocalDevelopers/motocal/commit/426c8a333da601b8624d49ace66dd4e6575f28ae)

> Doesn't stack with the Stamina support skills from Arulumaya (Summer), Izmir (Summer), Jeanne d'Arc and Sakura Kinomoto.
Whichever effect grants the highest ATK Up will take precedence.

 \-  https://gbf.wiki/True_Purity_Sunblade

Both the weapon and support ability are the "same".
moving the weapon's effect to `normalOtherKonshin` will create a 'conflict' with omegaKonshin (atma/ultima). also need to change code in `generateHaisuiData`.

**Comment by OrkunKocyigit** [comment]
Um it doesn't that way though because I substract skill from characters as they will be always added regardless.
totals[key]["normalOtherKonshin"] += Math.max(module.exports.calcHaisuiValue("normalSupportKonshin", "L", "1", totals["Djeeta"]["remainHP"]) - totals[key]["normalSupportKonshin"], 0);

So all 3 conditions:
- [ ] Character skill >= Weapon Skill then Weapon Skill - Character Skill =< 0 so it adds 0 (Aka only character value is used).
- [ ] Character skill < Weapon Skill then it adds difference of whatever is higher henceforth becomes non-issue again.
Like: If character gives 7 but weapon 9, it adds 2 so 7+2 = 9

For omega they stack multiplicative? If so that is also fixed able to multiply instead of sum.

**Comment by h-yasha** [comment]
I'm aware. It's just finding a solution where the weapon skills are also added to normalSupportKonshin to have the logic/data be more _organized_ is going to be much easier for others to go through and look much cleaner (by simply applying _x_ buff/skill/etc to only one name.

> For omega they stack multiplicative? If so that is also fixed able to multiply instead of sum. Though then current code also wrong as it just adds.

no they are additive, it's normal stamina.

**Comment by OrkunKocyigit** [comment]
@h-yasha @kei-gbf Do we have anything else that requires changes for this branch? If so would it be possible to describe it?

**Comment by kei-gbf** [comment]
@OrkunKocyigit 
- [ ] Update PR comment
  - [ ] summary of other updates

please paste this in PR comment.
```md
----
- weapon skill
  - [ ] [Cherubim lance](https://gbf.wiki/Cherubim%27s_Lance)
  - [ ] [True Purity Sunblade](https://gbf.wiki/True_Purity_Sunblade)
- stamina support ability
  - [ ] [Arulumaya (Summer)](https://gbf.wiki/Arulumaya_(Summer))
  - [ ] [Izmir (Summer)](https://gbf.wiki/Izmir (Summer))
  - [ ] [Jeanne d'Arc](https://gbf.wiki/https://gbf.wiki/Jeanne%20d'Arc)
  - [ ] [Sakura Kinomoto](https://gbf.wiki/https://gbf.wiki/Sakura%20Kinomoto)
```

----
## Checklist

list from wiki

- Cherubim lance
  - [ ] Has a Normal Stamina modifier, which is the same modifier as Weapon s  Fallen Sword, but cannot be boosted by Summon s  Hades.
  - [ ] 10% ATK at 100% HP and decreases to 1% ATK at 50% HP.
  - [ ] Does not stack with other Stamina-based character support skills. Only the highest ATK boost takes effect. This includes (but is not limited to):
- True Purity Sunblade
  - [ ] Has a Normal Stamina modifier, which is the same modifier as Weapon  Eden and Bow of Iliad, but not boosted by Summon  Zeus.
  - [ ] Check this in Emnity/Stamina Graph
    - [ ] 15% ATK Up at 100% HP;
    - [ ] 1.5% ATK Up at 50% HP;
    - [ ] 0% ATK Up below 50% HP.
  - [ ] Doesn't stack with the Stamina support skills from Arulumaya (Summer), Izmir (Summer), Jeanne d'Arc and Sakura Kinomoto.
  Whichever effect grants the highest ATK Up will take precedence.

File: src/global_logic.js [hunk 2]
```diff
@@ -2500,13 +2504,28 @@ module.exports.treatSupportAbility = function (totals, chara) {
...
+                                totals[key2]["normalSupportKonshin"] = Math.max(totals[key2]["normalSupportKonshin"], supportKonshinValue);
+                            }
+                        } else {
+                            totals[key]["normalSupportKonshin"] = Math.max(totals[key]["normalSupportKonshin"], supportKonshinValue);
+                        }
+                    }
+                    continue;
+                case "normalSupportKonshin_hpDebuff":
+                    {
                     let supportKonshinValue = module.exports.calcHaisuiValue("normalSupportKonshin", support.value, 1, totals[key]["remainHP"]);
```
Comment (kei-gbf) [inline src/global_logic.js]: this line need indent +4

File: src/global_logic.js [hunk 3]
```diff
@@ -2696,14 +2722,24 @@ module.exports.generateHaisuiData = function (res, arml, summon, prof, chara, st
                                 }
                                 omegaKonshinIncluded = true;
                             }
+                        } else if (stype == 'cherubimKonshin') {
+                            for (var l = 0; l < haisuiBuff.length; l++) {
+                                    let normalSupportKonshinValue = 0.01 * module.exports.calcHaisuiValue("normalSupportKonshin", "M", "1", 0.01 * (l + 1));
```
Comment (kei-gbf) [inline src/global_logic.js]: indent -4

**Comment by kei-gbf** [comment]
@OrkunKocyigit 

Can you test "Open Enmity Graph"
- Add stamina support chara
- Change Character's RemainHP ratio (or Party's remainHP in Player tab?)


stamina03

Info line show Izmir remain HP 1%
but seem the graph changes by Djeeta's remain HP. (based on real HP show Djeeta's)

**Comment by kei-gbf** [comment]
- [x] Remain HP 0%,50%,100% and ATK up 1%,1.5% ... 10%,15% ok

stamina01
stamina02

**Comment by kei-gbf** [comment]
Checked
- [x] Normal Stamina modifier
- [x] but cannot be boosted by Summon
- [ ] just minor indent fix

@OrkunKocyigit 

The Enmity Graph was ok, my misunderstood.

**Comment by OrkunKocyigit** [comment]
@kei-gbf Would it be okay with you if I add this rule to eslint and run eslint --fix. I think it might update a lot of files but would fix all of our intent issues.
"indent": ["error", 4]

**Comment by kei-gbf** [comment]
@OrkunKocyigit 
eslint --fix  no for now.
that would make a lot conflicts. there are a lot PR.
editorconfig is a partial solution for this.

I am conservative to legacy code. it's a fear to change without tests widely.

----
> provides function wrapper for chart.js

thx update PR comment, I had not catch this.

File: src/global_logic.js (hunk 2 above)
Comment (OrkunKocyigit) [reply src/global_logic.js]: Fixed

File: src/global_logic.js (hunk 3 above)
Comment (OrkunKocyigit) [reply src/global_logic.js]: Fixed

**Comment by OrkunKocyigit** [comment]
Fixed indent errors, I'll create another PR for fixing ESLint

File: src/global_logic.js (hunk 2 above)
Comment (kei-gbf) [reply src/global_logic.js]: !? no I thought it was only single line needed indent spaces +4
now "continue" has different indent level ?

if use block scope for "case"
there is a style swith and case in the same indent level.
but I did not mention that.

**Comment by kei-gbf** [comment]
- "let" line had different indent level in
  - case "normalSupportKonshin":
  - case "normalSupportKonshin_hpDebuff":

```diff
                case "normalSupportKonshin":
                    {
                        let supportKonshinValue = module.exports.calcHaisuiValue("normalSupportKonshin", support.value, 1, totals[key]["remainHP"]);
                        if (totals[key].isConsideredInAverage) {
                            for (var key2 in totals) {
                                totals[key2]["normalSupportKonshin"] = Math.max(totals[key2]["normalSupportKonshin"], supportKonshinValue);
                            }
                        } else {
                            totals[key]["normalSupportKonshin"] = Math.max(totals[key]["normalSupportKonshin"], supportKonshinValue);
                        }
                    }
                    continue;
                case "normalSupportKonshin_hpDebuff":
                    {
-                    let supportKonshinValue = module.exports.calcHaisuiValue("normalSupportKonshin", support.value, 1, totals[key]["remainHP"]);
+                        let supportKonshinValue = module.exports.calcHaisuiValue("normalSupportKonshin", support.value, 1, totals[key]["remainHP"]);
                        if (totals[key].isConsideredInAverage) {
                            for (var key2 in totals) {
                                totals[key2]["normalSupportKonshin"] = Math.max(totals[key2]["normalSupportKonshin"], supportKonshinValue);
                            }
                        } else {
                            totals[key]["normalSupportKonshin"] = Math.max(totals[key]["normalSupportKonshin"], supportKonshinValue);
                        }
                        totals[key]["HPdebuff"] += support.hpDebuff;
                    }
                    continue;
```

**Comment by kei-gbf** [comment]
If the stamina logic are same, the diff was only hp debuff

```js
case "normalSupportKonshin_hpDebuff":
    totals[key]["HPdebuff"] += support.hpDebuff;
    /* FALLTHROUGH */
case "normalSupportKonshin":
    :
```

**Comment by OrkunKocyigit** [comment]
> If the stamina logic are same, the diff was only hp debuff

I kinda find fall through a bit risky as if someone moves the code in places it tends to break but I can remove continue block

File: src/global_logic.js (hunk 2 above)
Comment (OrkunKocyigit) [reply src/global_logic.js]: It was because cases were in it is static scope. It has been removed in my latest commit

**Comment by kei-gbf** [comment]
switch/fall-through is an idiom, eslint can pass this with the comment line.

https://eslint.org/docs/rules/no-fallthrough

I suggest add hpDebuff in normalSupportKonshin_hpDebuff
normalSupportKonshin no need to calc debuff zero. it make mislead for readers.

**Comment by OrkunKocyigit** [comment]
> 
> 
> switch/fall-through is an idiom, eslint can pass this with the comment line.
> 
> https://eslint.org/docs/rules/no-fallthrough
> 
> I suggest add hpDebuff in normalSupportKonshin_hpDebuff
> normalSupportKonshin no need to calc debuff zero. it make mislead for readers.

Updated to fall through.

**Comment by kei-gbf** [comment]
@OrkunKocyigit 

I agree about the risk of fall through.
ESLint will report about the silence fall through you worry.

fall through was just temporary solution for this clone code.
implement the common part as separate function is better. (except passing many parameters)

----
> provides function wrapper for chart.js global_logic.js/recalcNormalSupportKonshin

thx now I was looking for this info.

> recalcNormalSupportKonshin

pls add comment "fall through" for ESLint.
(Although current ESLint may not check this rule)

**Comment by OrkunKocyigit** [comment]
@kei-gbf 
> pls add comment "fall through" for ESLint.
> (Although current ESLint may not check this rule)
Added.

**Comment by kei-gbf** [comment]
@OrkunKocyigit 

ok I confirmed:
- fall through comment
  - [x] treatSupportAbility
  - [x] recalcNormalSupportKonshin 

----
it's really minor stuff. order of cases. hpDebuff first

```js
case "normalSupportKonshin_hpDebuff":
    // fall through
case "normalSupportKonshin":
```

in the fall through case, the order can explain to readers
`normalSupportKonshin_hpDebuff` includes `normalSupportKonshin` 
nothing different to result though. (it's for code-reading)

----
I am now testing Average include off cases.
I found the else case had similar
```js
// Calculate yourself only if you do not put it in the average
```

I do not understand the details of average include,
just looks ok as follow other code.

**Comment by kei-gbf** [comment]
> I'll create another PR for fixing ESLint conf and making it more useful

I had tried eslint --fix  with some react related plugins.
that changed a lot lines.

Automatic convert is ok itself, It does not mean I don't trust eslint --fix
I worry point is manually change by solve the merge conflict.
When code were changes widely (e.g indent only changes etc...)
that make conflict to other branches. others may need to solve it manually.

I feel we need schedule management for this task.
if all collaborators could restart PR from new branch, it's no problem.
currently, a lot PRs, that may make a lot merge conflict.

**Comment by OrkunKocyigit** [comment]
> 
> 
> > I'll create another PR for fixing ESLint conf and making it more useful
> 
> I had tried eslint --fix with some react related plugins.
> that changed a lot lines.
> 
> Automatic convert is ok itself, It does not mean I don't trust eslint --fix
> I worry point is manually change by solve the merge conflict.
> When code were changes widely (e.g indent only changes etc...)
> that make conflict to other branches. others may need to solve it manually.
> 
> I feel we need schedule management for this task.
> if all collaborators could restart PR from new branch, it's no problem.
> currently, a lot PRs, that may make a lot merge conflict.
> 
> ADD: Another point currently project missing is coverage tests.
> to check the wide changes.

That is basically where I am at. Our lint setup missing a lot of good configs and missing a lot of profiles for frameworks we use, React comes as example. I tried to setup all that in my local branch but I got 498 errors that needed to be fixed manually.

I'll agree that full regression set should take priority over that. Once that is done we can talk about setting a temporary freeze to activate those format changes.

**Comment by kei-gbf** [comment]
Question, [Vira (Summer)](https://gbf.wiki/Vira_(Summer)) skill2 is also same stamina?
can use that to simulate the skill2?
(not change request, it is not support ability)

----
- [x] just note, recalcNormalSupportKonshin is also clone code of recalcCharaHaisui 
  - I have an idea to solve but I don't want to change recalcCharaHaisui now
   this can be future topic when add many recalc* functions. the workflow should be with tests
    - unit test for current recalcCharaHaisui
    - change recalcCharaHaisui implementation (fix the clone code issue)
    - run tests to confirm to get the same result

----
thank you @OrkunKocyigit and @h-yasha for the first review.

BTW, do you join Motocal developers group? (I will ask someone who has the permission)
There are a lot PR need reviews, need help, and I want time for my own PR too.

**Comment by OrkunKocyigit** [comment]
> BTW, do you join Motocal developers group? (I will ask someone who has the permission)
> There are a lot PR need reviews, need help, and I want time for my own PR too.

Sure I'll join if given opportunity.

**Comment by h-yasha** [comment]
> Question, [Vira (Summer)](https://gbf.wiki/Vira_(Summer)) skill2 is also same stamina?
> can use that to simulate the skill2?
> (not change request, it is not support ability)

I don't have verification for this, but Strength/Jammed are _probably (as they're unique, both are)_ not the same as this support ability/MH skill.

**Comment by kei-gbf** [comment]
need to wait a response, seem they may be busy/non active.

----
from wiki link

max value is 15% same
min value was 5% != 1%

I had not seen the actual damage log of verification, just the info was a bit different.
and for now, as I search, no explicit info about stack or not.

**Comment by kei-gbf** [comment]
渾身 メイン武器効果＆サポアビ実装のレビュー完了しました。

- 武器 (メイン装備時)
  - ランス･オブ･ケルヴィム
  - 真・道天浄土
- キャラ (サポアビ)
  - 水着イシュミール
  - 水着アルルメイヤ
  - ジャンヌダルク
  - 木之元桜

詳細はPRのコメント参照

----
Review done
Add stamina main weapon skill/character support ability.
Details at the first comment https://github.com/MotocalDevelopers/motocal/pull/237#issue-285873228

**Comment by OrkunKocyigit** [comment]
@kei-gbf 
> need to wait a response, seem they may be busy/non active.
> 
> from wiki link
> 
> max value is 15% same
> min value was 5% != 1%
> 
> I had not seen the actual damage log of verification, just the info was a bit different.
> and for now, as I search, no explicit info about stack or not.

I tested on my gf account, it looks like unique multiplier compared to this one that is normal multiplier.

**Pull request merged**